event-sourcing/event-store-jpa/src/main/java/org/apache/james/eventsourcing/eventstore/jpa/JPAEventStore.java

```diff
@@ -0,0 +1,176 @@
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.EntityTransaction;
+import jakarta.persistence.PersistenceException;
+import jakarta.persistence.PersistenceUnit;
+
+import org.apache.james.backends.jpa.EntityManagerUtils;
+import org.apache.james.eventsourcing.AggregateId;
+import org.apache.james.eventsourcing.Event;
+import org.apache.james.eventsourcing.eventstore.EventStore;
+import org.apache.james.eventsourcing.eventstore.EventStoreFailedException;
+import org.apache.james.eventsourcing.eventstore.History;
+import org.apache.james.eventsourcing.eventstore.JsonEventSerializer;
+import org.apache.james.eventsourcing.eventstore.jpa.model.JPAEvent;
+import org.reactivestreams.Publisher;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+import scala.collection.Iterable;
+
+
+public class JPAEventStore implements EventStore {
+
+    /**
+     * The entity manager to access the database.
+     */
+    private EntityManagerFactory entityManagerFactory;
+
+    /**
+     * The JSON serializer to serialize the event data.
+     */
+    private JsonEventSerializer jsonEventSerializer;
+
+    /**
+     * Set the serializer to use.
+     */
+    @Inject
+    public void setJsonEventSerializer(JsonEventSerializer jsonEventSerializer) {
+        this.jsonEventSerializer = jsonEventSerializer;
+    }
+
+    /**
+     * Set the entity manager to use.
+     */
+    @Inject
+    @PersistenceUnit(unitName = "James")
+    public void setEntityManagerFactory(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
+    }
+
+    @Override
+    public Publisher<Void> appendAll(Iterable<Event> events) {
+        if (events.isEmpty()) {
+            return Mono.empty();
+        }
+        Preconditions.checkArgument(Event.belongsToSameAggregate(events));
+        AggregateId aggregateId = events.head().getAggregateId();
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        final EntityTransaction transaction = entityManager.getTransaction();
+        try {
+            transaction.begin();
+            events.foreach(e -> {
+                try {
+                    JPAEvent jpaEvent = new JPAEvent(aggregateId, e.eventId(), jsonEventSerializer.serialize(e));
+                    entityManager.persist(jpaEvent);
+                } catch (JsonProcessingException ex) {
+                    throw new RuntimeException(ex);
+                }
+                return e;
+            });
+            transaction.commit();
+        } catch (PersistenceException e) {
+            if (transaction.isActive()) {
+                transaction.rollback();
+            }
+            EventStoreFailedException esfe = new EventStoreFailedException("Unable to add events");
+            esfe.initCause(e);
+            throw esfe;
```

Comment:
```suggestion
            throw  new EventStoreFailedException("Unable to add events", e);
```

Mutability can be avoided

Reply:
That's what I intended to do, but this exception doesn't support a cause parameter... I left the scala api part alone for now.